### PR TITLE
V2/master code specification and remove some useless code

### PR DESCRIPTION
**Pull request by bxlxx**
> Thank you for contributing to Coraza WAF, your effort is greatly appreciated
> Before submitting check if what you want to add to `coraza-waf` list meets [quality standards](https://github.com/jptosso/coraza-waf/blob/master/CONTRIBUTING.md#quality-standards) before sending pull request. Thanks!

**Note**: _that go.mod and go.sum can only be modified for tested dependency updates or justified new features._

**Make sure that you've checked the boxes below before you submit PR:**

- [x] My code includes positive and negative tests.
- [x] I have an appropriate description with correct grammar.
- [x] I have read [Contribution guidelines](https://github.com/jptosso/coraza-waf/blob/master/CONTRIBUTING.md#contribution-guidelines), [maintainers note](https://github.com/jptosso/coraza-waf/blob/master/CONTRIBUTING.md#maintainers) and [Quality standard](https://github.com/jptosso/coraza-waf/blob/master/CONTRIBUTING.md#quality-standards).

Thanks for your PR :heart:

### CODE REVIEW

File: body_buffer.go [hunk 1]
```diff
@@ -23,7 +23,6 @@ import (
 // BodyBuffer is used to read RequestBody and ResponseBody objects
 // It will handle memory usage for buffering and processing
 type BodyBuffer struct {
-	io.Writer   // OK?
```
Comment (jptosso) [inline body_buffer.go:26]: It was left for developers to know that BodyBuffer implements io.Writer, so they can
```go
io.Copy(bodybuffer, someReader)
```
What do you think @fzipi ?

File: rule.go [hunk 2]
```diff
@@ -225,6 +225,7 @@ type Rule struct {
 func (r *Rule) Evaluate(tx *Transaction) []MatchData {
 	if r.Capture {
 		tx.Capture = true
+		defer tx.resetCaptures()
```
Comment (jptosso) [inline rule.go:228]: I think there is a reason why I added r.Capture validation twice, resetCaptures is CPU expensive, so I try to avoid it as much as possible. We must check if that is the case

File: rule_match.go [hunk 3]
```diff
@@ -35,6 +35,10 @@ type MatchData struct {
 	Value string
 }
 
+func (m MatchData) isNil() bool {
```
Comment (jptosso) [inline rule_match.go:38]: It would be cool to add some documentation on why use isNil

File: rulegroup.go [hunk 4]
```diff
@@ -171,12 +166,6 @@ RulesLoop:
 		tx.GetCollection(variables.MatchedVars).Reset()
 		tx.GetCollection(variables.MatchedVarsNames).Reset()
 
-		txr := tx.GetCollection(variables.Rule)
```
Comment (jptosso) [inline rulegroup.go:174]: Why did you remove this, is it duplicated? Rules can query their metadata, like

```
SecRule RULE:id "15" "id:15"
```

this should match

File: seclang/directives.go [hunk 5]
```diff
@@ -121,7 +121,7 @@ func directiveSecRule(w *coraza.Waf, opts string) error {
 }
 
 func directiveSecResponseBodyAccess(w *coraza.Waf, opts string) error {
-	w.ResponseBodyAccess = (strings.ToLower(opts) == "on")
```
Comment (jptosso) [inline seclang/directives.go:124]: There is a helper that should be used: ```parseBoolean(opts)```, I forgot to use it

File: seclang/directives.go [hunk 6]
```diff
@@ -203,7 +203,7 @@ func directiveSecResponseBodyLimit(w *coraza.Waf, opts string) error {
 }
 
 func directiveSecRequestBodyLimitAction(w *coraza.Waf, opts string) error {
-	w.RejectOnRequestBodyLimit = (strings.ToLower(opts) == "reject")
+	w.RejectOnRequestBodyLimit = strings.ToLower(opts) == "reject"
```
Comment (jptosso) [inline seclang/directives.go:206]: @fzipi should this be transformed into a type? there are only two possibilities

File: types/variables/variables.go [hunk 7]
```diff
@@ -25,8 +25,8 @@ import (
 // This file repeats the same content many times in order to make access
 // efficient for seclang and transactions
 
-// VariablesCount contains the number of existing variables
-const VariablesCount = 87
+// Count contains the number of existing variables
+const Count = 87
```
Comment (jptosso) [inline types/variables/variables.go:29]: The reason I used VariablesCount instead of Count is because Count can be easily mistaken as a valid seclang variable, what do you think?

**Review by jptosso — COMMENTED** [review]
Everything is fine and works but I have some observations that I would like to discuss

**Comment by jptosso** [comment]
By the way, the coresetrule compatibility test is not working because of recent CRS changes, it should be fixed soon.

File: types/variables/variables.go (hunk 7 above)
Comment (bxlxx) [reply types/variables/variables.go:29]: I think the ```Count``` is under the variables package, it would be cumbersome to use ```VariablesCount```.

File: rule.go (hunk 2 above)
Comment (bxlxx) [reply rule.go:228]: If defer is placed if a for loop, it may cause multiple executions after the function ends.

File: rule_match.go (hunk 3 above)
Comment (bxlxx) [reply rule_match.go:38]: That's ok. I'll add it.

File: types/variables/variables.go (hunk 7 above)
Comment (jptosso) [reply types/variables/variables.go:29]: I agree with you, what about moving VariablesCount to the types package? It still bugs me under variables

File: rulegroup.go (hunk 4 above)
Comment (bxlxx) [reply rulegroup.go:174]: Yeah, it's duplicated. The code for the same function as follows is used in the subsequent [```r.Evaluate(tx)```](https://github.com/jptosso/coraza-waf/blob/56ff1874903311ba7f497f1fec2c9aaac5ab68e5/rulegroup.go#L180) function. [the same code link](https://github.com/jptosso/coraza-waf/blob/56ff1874903311ba7f497f1fec2c9aaac5ab68e5/rule.go#L249):
```
tx.GetCollection(variables.Rule).SetData(map[string][]string{
	"id":       {strconv.Itoa(rid)},
	"msg":      {r.Msg.String()},
	"rev":      {r.Rev},
	"logdata":  {r.LogData.String()},
	"severity": {r.Severity.String()},
})
```

File: seclang/directives.go (hunk 5 above)
Comment (bxlxx) [reply seclang/directives.go:124]: I'll add it.

File: rulegroup.go (hunk 4 above)
Comment (jptosso) [reply rulegroup.go:174]: You are right, thank you for noticing

File: types/variables/variables.go (hunk 7 above)
Comment (bxlxx) [reply types/variables/variables.go:29]: It's ok with moving the ```VariablesCount``` to the types package so that it will not appear wordy.

Comment (bxlxx) [reply types/variables/variables.go:29]: I discard the change here, I'll revert.

File: rule.go (hunk 2 above)
Comment (jptosso) [reply rule.go:228]: You are right, I think it could be a smarter function like 

```go
defer func (){
 if thereWereMatches { 
   tx.resetCaptures()
  }
}
```

It's a super memory expensive operation.
But let's leave it like this for now.

File: body_buffer.go (hunk 1 above)
Comment (jptosso) [reply body_buffer.go:26]: Approved, please add to the documentation that it implements io.Writer
